### ISSUE TITLE
How the hell does FastAPI's module system work?

### ISSUE THREAD

**Issue by nocturn9x** [tiangolo/fastapi#3375]
### First check

* [X] I added a very descriptive title to this issue.
* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [X] I already read and followed all the tutorial in the docs and didn't find an answer.
* [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [X] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

Ok, so, I'll make this super easy to understand: please, please, PLEASE review FastAPI's documentation, because it feels like reading a children's book. I honestly feel like I'm reading something aimed towards mentally challenged people and I every time I open up the documentation to find a solution to something my brain is like "can you just get to the damn point?". I've been developing software for over 6 years and I don't an explanation of what a coroutine is, how dependency injection works or what HTTP is. Frankly I'm puzzled as to why a framework aimed at developers which should supposedly know what they're doing includes all of this kindergarten-level stuff. Now, to get to my question: How the hell do I make my FastAPI application modular? I have over 3500 lines of code I need to split into several modules and the docs on the topic are just retarded, sorry there's no other way to put it. I get the "being friendly" to beginners thing, but this is beyond ridicolous

**Comment by Kludex** [comment]
@nocturn9x That tone is not appreciated in this community. That's not how you ask for help or give feedback.

**Comment by chbndrhnns** [comment]
@nocturn9x Now that it's pretty clear that you don't like the documentation, we still do not know what exact problem you are facing with splitting your code up into modules.
Do you have any more to say on this actual problem?

**Comment by nocturn9x** [comment]
@chbndrhnns The problem is quite simple, the solution probably is too but the documentation does its best to dilute it into a bag of nonsense most experienced developers do not care about. I don't have a specific problem because I couldn't even figure out how use the `APIRouter` class properly, there's too much confusing wording in between to make some sense out of it

**Comment by nocturn9x** [comment]
> @nocturn9x That tone is not appreciated in this community. That's not how you ask for help or give feedback.

@Kludex Sorry for my tone, but I need this framework to write actual production software and not toys like the docs suggest, and the way that whole website is laid out makes me go nut

**Comment by acutaia** [comment]
@nocturn9x  If you want an example of a real app, you could look here in the [source code ](https://github.com/nsidnev/fastapi-realworld-example-app). Regarding the docs, I think are perfect, cause everyone has a different background, and also I think that this is a framework made for everyone not only for experts. If you don't like them, you could rewrite them and open a pull request to share with the community

**Comment by agropper** [comment]
I, for one, really appreciate the support for beginners and the welcoming
and inclusive response to issues.

**Comment by acidjunk** [comment]
Even after the intro was rewritten (and the tone is a lot nicer) it doesn't show what the actual problem is: This could off course just be lost somewhere in the rant about the docs.

We (SURF) manage a couple of large codebases with FastAPI and all the normal python module stuff just works. There are several layers you can use the make a very modular app. Splitting models, api routes, frequently used business logic etc.

I did some work on a boilerplate (it's not finished yet) that shows how you could split code in a DB heavy app: https://github.com/acidjunk/fastapi-postgres-boilerplate

But that being said; I also manage quite some apps that don't even use a DB: so I think making it modular is something that you determine on a per project basis. FastAPI isn't as opinionated as Django in that sense.

**Comment by frankie567** [comment]
There is a WHOLE PAGE dedicated to modularity and how to structure a FastAPI app: https://fastapi.tiangolo.com/tutorial/bigger-applications

It's crystal clear how to split your app in several routers and how to include them in your main app.

**Comment by ArcLightSlavik** [comment]
>I've been developing software for over 6 years and I don't need nor want an explanation of what a coroutine is, how dependency injection works or what HTTP is.

One of good skills of **experienced** programmers is ability to filter out the junk in a tutorial and find the parts **they** require.  Say scrolling through a tutorial for AWS deployment. I do not care about creation of my profile or the basic app they deploy on it. I already did that, but for example on a front-end CSS tutorial every single sentence is important because I don't know anything about css.

**Comment by dithmer** [comment]
Such a grumpy request doesn't deserve any kind of effort put into good answers, like those you got already. There should be nothing more said, but a big thank you by you, to those nice people.

**Comment by josnin** [comment]
I really appreciate how the documentation was written, one of the many reason why i started using it

**Comment by borland667** [comment]
I completely agree with @dithmer. I wouldn't expect anyone to reply to such grumpy request. Documentation is absolutely fantastic and as an experienced engineer one should be able to filter out anything not important or not required at the time and focus on what you need.

**Comment by thaodt** [comment]
Hi there, totally agreed with FastAPI supporters / contributors, specially author of this framework.
Almost my projects in the past were Flask, and I also joined some Django projects at work.

I actually appreciate FastAPI community because their docs is pretty cool and easy getting started. 
What I do care about when starting a project in the real world (production things & give proposal to my boss) are how project is structured - big application. And FastAPI did it - https://fastapi.tiangolo.com/tutorial/bigger-applications

[Deployment](https://fastapi.tiangolo.com/deployment/) ? yes, they have it also.

And the author also give us [his projects skeleton](https://fastapi.tiangolo.com/deployment/docker/) if you do read the docs thoroughly 

And if you really need some projects out there, check [this](https://fastapi.tiangolo.com/external-links/) out.

For me, Flask docs is really poor. i don't mean to criticize. This is my opinion for the old Flask versions, recently its released new version, still didnt have time to check it out.

How about Django? Their docs is really big and hard to getting started for beginners to understand its features completely. 

But these are my opinions. Actually when you want to understand a new thing completely, just go to their source code and read it. You may learn something new, cool and free your mind.

**Comment by fabioueno** [comment]
@nocturn9x, the documentation isn't supposed to serve the expert developers, but also more beginner ones. With that said, there are more advanced topics, and, as stated by @frankie567, there's a page on the modularity of bigger applications:

> There is a WHOLE PAGE dedicated to modularity and how to structure a FastAPI app: https://fastapi.tiangolo.com/tutorial/bigger-applications

Does this resolve your issue? 🤔

I believe restructuring the whole documentation is an effort out of question, especially if it means making it more complex for more novice developers.
